{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import re\n",
    "from Bio.Seq import Seq\n",
    "from Bio import Restriction\n",
    "import random\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%% load mouse 3' utrs from biomart - GRCm38.p6\n",
    "\n",
    "input_file = './Datasets/mouse_3utrs.txt'\n",
    "\n",
    "#fasta_sequences = SeqIO.parse(open(input_file),'fasta')\n",
    "\n",
    "utrsdict={rec.id : str(rec.seq) for rec in SeqIO.parse(input_file, \"fasta\")}\n",
    "#for fasta in fasta_sequences:\n",
    "#    name, sequence = fasta.id, str(fasta.seq)\n",
    "#    utrsdict[name]=sequence\n",
    "\n",
    "utrsdictdf=pd.DataFrame.from_dict(data=utrsdict, orient='index').iloc[:,0]\n",
    "\n",
    "genenames={rec.id : rec.id.split('|')[2] for rec in SeqIO.parse(input_file, \"fasta\")}\n",
    "#for key in utrsdict.keys():\n",
    "#    genenames[key]=key.split('|')[2]\n",
    "\n",
    "genenamesdf=pd.DataFrame.from_dict(data=genenames, orient='index').iloc[:,0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "55486"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genenamesdf.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "genesutrs={}\n",
    "for vals in genenamesdf.unique():\n",
    "    genesutrs[vals]=list(genenamesdf[genenamesdf==vals].index)\n",
    "\n",
    "genesutrsdf=pd.DataFrame.from_dict(data=genesutrs, orient='index').iloc[:,0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "genesutrsdf.to_pickle('./dataframes/genesutrsdf.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Wnt9a         ENSMUSG00000000126|ENSMUST00000000128|Wnt9a|11...\n",
       "Cntn1         ENSMUSG00000055022|ENSMUST00000000109|Cntn1|15...\n",
       "Ngfr          ENSMUSG00000000120|ENSMUST00000000122|Ngfr|11|...\n",
       "Scmh1         ENSMUSG00000000085|ENSMUST00000000087|Scmh1|4|...\n",
       "Wnt3          ENSMUSG00000000125|ENSMUST00000000127|Wnt3|11|...\n",
       "                                    ...                        \n",
       "AL772407.1    ENSMUSG00000118517|ENSMUST00000239054|AL772407...\n",
       "AC087801.2    ENSMUSG00000118583|ENSMUST00000239183|AC087801...\n",
       "AC141882.1    ENSMUSG00000118503|ENSMUST00000239096|AC141882...\n",
       "AC122861.1    ENSMUSG00000118610|ENSMUST00000239201|AC122861...\n",
       "AC124742.1    ENSMUSG00000118552|ENSMUST00000238938|AC124742...\n",
       "Name: 0, Length: 55486, dtype: object"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genesutrsdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "utrsdictdf=pd.Series({rec.id.split('|')[2] : str(rec.seq) for rec in SeqIO.parse(\"./Datasets/mouse_3utrs.txt\", \"fasta\")})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%% functions\n",
    "\n",
    "def get_wt_utrs(genelist):\n",
    "    genelistdf=pd.DataFrame()\n",
    "    for gene in genelist:\n",
    "        try:\n",
    "            genedf=utrsdictdf.loc[genesutrs[gene]]\n",
    "            \n",
    "            genedf.drop(genedf[genedf=='Sequenceunavailable'].index, inplace=True)\n",
    "            genedf=genedf.to_frame()\n",
    "            genedf.loc[:,'gene_name']=gene\n",
    "            genedf.columns=['sequence','gene_name']\n",
    "            \n",
    "            for i in genedf.index:\n",
    "                for j in genedf.index:\n",
    "                    if (j!=i) & bool(genedf.loc[j,'sequence'].find(genedf.loc[i,'sequence'])>-1):\n",
    "                        genedf.drop(i, inplace=True)\n",
    "                        break\n",
    "            genelistdf=genelistdf.append(genedf)\n",
    "        except:\n",
    "            pass\n",
    "    genelistdf['seqlength']=genelistdf.sequence.apply(lambda x: len(x))\n",
    "    genelistdf['paposition']=genelistdf.sequence.apply(lambda seq: len(seq)-50+re.search('A[AT]TAAA', seq[-50:]).span()[0] if not\n",
    "                              re.search('A[AT]TAAA', seq[-50:])==None else -1)\n",
    "    genelistdf['sequence_minuspa']=genelistdf.apply(lambda x: x.loc['sequence']\n",
    "               [:x.loc['paposition']] if x.loc['paposition']>0 else x.loc['sequence'], axis=1)\n",
    "    return genelistdf\n",
    "            \n",
    "def complists(list1, list2):\n",
    "    listout=[]\n",
    "    for i in list1:\n",
    "        if i in list2:\n",
    "            listout+=[i]\n",
    "    print(len(listout))\n",
    "    return listout\n",
    "\n",
    "def make_scanning(sq, increment):\n",
    "    varseqs=[]\n",
    "    remaining=len(sq)\n",
    "    startpos=0\n",
    "    while remaining>=150:\n",
    "        varseqs.append(sq[startpos:startpos+150])\n",
    "        startpos+=increment\n",
    "        remaining-=increment\n",
    "    return varseqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.isdir('./dataframes/design')==False:\n",
    "    os.mkdir('./dataframes/design')\n",
    "\n",
    "posgenes=['Actb','Arc','Camk2a','Ranbp1','Bdnf','Map2','Shank1','Dlg4','Map1a','Cplx2',\n",
    "          'Calm1','Grin1','Nrgn','Ranbp1']\n",
    "\n",
    "posgenesdf=get_wt_utrs(posgenes) # 17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16, 5)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "posgenesdf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17\n",
      "4\n",
      "127\n",
      "6\n"
     ]
    }
   ],
   "source": [
    "#%% pick contexts\n",
    "\n",
    "'''\n",
    "Middleton et al., 2019:\n",
    "    deDent - significantly enriched in neurites\n",
    "    consDent - found in 90% of the cells in the neurite (and also in the soma potentially)\n",
    "    isoDent - genes with differential localization of isoforms\n",
    "    \"We found 1863 RNAs in at least 90% of the dendrite samples, which included \n",
    "    well-characterized localized RNAs such as Actb, Bdnf, Calm1, Dlg4, Grin1, \n",
    "    and Map2. \"\n",
    "\n",
    "    \n",
    "Zappulo et al., 2017:\n",
    "    \"We identified 1,292 transcripts enriched in neurites by at least 2-fold when \n",
    "    compared with the soma (P-values < 0.05).\"\n",
    "    positive controls: \"We found transcripts known to be preferentially localized, \n",
    "    such as syntaxin-3 (Stx3)22, glutamate receptor-1 (Gria1)5, calcium channel \n",
    "    Ryr2 5, inositol 1,4,5-trisphosphate receptor type 1 (Itpr1)5, neuregulins \n",
    "    (Nrg1 23 and Nrg2 24), voltage-dependent L-type calcium channel subunit α-1D \n",
    "    (Cacna1d)5, ephrin type-A receptor 2 (Epha2)25, unconventional myosin-Ic \n",
    "    (Myo1c)26, low-density lipoprotein receptor adapter protein 1 (Ldlrap1)27, \n",
    "    vang-like protein (Vangl)28, and transcripts encoding mitochondrial proteins6,\n",
    "    10, consistent with previous works (see also Supplementary Data 5). \"\n",
    "    \n",
    "Taliaferro et al., 2016:\n",
    "    \"Several genes with known projection-enriched RNA localization patterns, \n",
    "    including β-actin (Actb), neurogranin (Nrgn), and Ranbp1, were identified \n",
    "    as neurite-enriched in both CAD and N2A cells\"\n",
    "    isoform spec: \"Previously, differential localization of alternative mRNA isoforms \n",
    "    has been observed in a few cases (An et al., 2008, Buckley et al., 2011, Harrison \n",
    "    et al., 2014, Whittaker et al., 1999).\"\n",
    "    \"Using more stringent criteria, we identified a confident set of 195 localized ALEs \n",
    "    and 96 localized tandem UTRs (Figure S1O). \"\n",
    "    \"We identified 421 distal ALE isoforms that were preferentially localized to \n",
    "    neurites, with criteria including ΔΨ ≥ 0.1 in both CAD and N2A cells (Table S2).\"\n",
    "\n",
    "Ciolli Mattioli et al., 2019:\n",
    "    alternative utrs\n",
    "    TableS2.xlsx\n",
    "\n",
    "Additional established neurite-localizing transcripts:\n",
    "    Arc\n",
    "    Camk2a\n",
    "    Ranbp1\n",
    "    Bdnf\n",
    "    Map2\n",
    "    Shank1\n",
    "    Dlg4\n",
    "    Map1a\n",
    "    Cplx2\n",
    "    \n",
    "'''\n",
    "\n",
    "if os.path.isdir('./dataframes/design')==False:\n",
    "    os.mkdir('./dataframes/design')\n",
    "\n",
    "posgenes=['Actb','Arc','Camk2a','Ranbp1','Bdnf','Map2','Shank1','Dlg4','Map1a','Cplx2',\n",
    "          'Calm1','Grin1','Nrgn','Ranbp1']\n",
    "\n",
    "posgenesdf=get_wt_utrs(posgenes) # 17\n",
    "\n",
    "posgenesdf.to_pickle('./dataframes/design/posgenesdf.pkl')\n",
    "\n",
    "\n",
    "# 40 top\n",
    "\n",
    "middleton_comp=pd.read_excel('./Datasets/MiddletonKimBMCBiol2019/12915_2019_630_MOESM6_ESM.xlsx')\n",
    "\n",
    "top40=middleton_comp.GeneName.head(n=40).values\n",
    "\n",
    "top40df=get_wt_utrs(top40) # 52\n",
    "\n",
    "top40df.to_pickle('./dataframes/design/top40df.pkl')\n",
    "\n",
    "\n",
    "# Middleton\n",
    "\n",
    "deDent=pd.read_excel('./Datasets/MiddletonKimBMCBiol2019/12915_2019_630_MOESM6_ESM.xlsx', \n",
    "                     sheet_name='TableS5-deDend genes')\n",
    "\n",
    "deDentlist=deDent.iloc[:,0].values\n",
    "\n",
    "deDend_df=get_wt_utrs(deDentlist) # 347\n",
    "\n",
    "deDend_df.to_pickle('./dataframes/design/deDend_df.pkl')\n",
    "\n",
    "\n",
    "consDent=pd.read_excel('./Datasets/MiddletonKimBMCBiol2019/12915_2019_630_MOESM6_ESM.xlsx', \n",
    "                     sheet_name='TableS6-consDend genes')\n",
    "\n",
    "consDentlist=consDent.iloc[:,0].values\n",
    "\n",
    "consDend_df=get_wt_utrs(consDentlist) # 2790\n",
    "\n",
    "consDend_df.to_pickle('./dataframes/design/consDend_df.pkl')\n",
    "\n",
    "\n",
    "isoDent=pd.read_excel('./Datasets/MiddletonKimBMCBiol2019/12915_2019_630_MOESM6_ESM.xlsx', \n",
    "                     sheet_name='TableS7-isoDend genes')\n",
    "\n",
    "isoDentlist=isoDent.GeneName.values\n",
    "\n",
    "isoDend_df=get_wt_utrs(isoDentlist) # 533\n",
    "\n",
    "isoDend_df.to_pickle('./dataframes/design/isoDend_df.pkl')\n",
    "\n",
    "\n",
    "#Ciolli Matiolli Isoforms\n",
    "\n",
    "ciolli=pd.read_excel('./Datasets/CiolliMattioliChekulaevaNAR2018_gky1270_supplemental_files/TableS2.xlsx', header=2)\n",
    "\n",
    "ciolli['gene_name']=ciolli.Names.apply(lambda x: x.split('|')[1])\n",
    "\n",
    "ciolligenes=ciolli.gene_name.unique()\n",
    "\n",
    "asoverlap=complists(ciolligenes, isoDentlist)\n",
    "\n",
    "asoverlap_df=get_wt_utrs(asoverlap) # 33\n",
    "\n",
    "\n",
    "# Zappulo\n",
    "\n",
    "zappulo3=pd.read_excel('./Datasets/ZappuloChekulaevaNatCommun2017/SuppData2.xlsx', header=0)\n",
    "zappulo6=pd.read_excel('./Datasets/ZappuloChekulaevaNatCommun2017/41467_2017_690_MOESM6_ESM.xlsx', header=1)\n",
    "\n",
    "zappulo_dend=zappulo3[(zappulo3['rzRNA_log2FC_Neurite_Soma']>2)&(zappulo3.rzRNA_padj_Neurite_Soma<0.01)].gene_name.unique()\n",
    "\n",
    "zappulo_dend_df=get_wt_utrs(zappulo_dend) # 127\n",
    "\n",
    "zappulo_dend_df.to_pickle('./dataframes/design/zappulo_dend_df.pkl')\n",
    "\n",
    "\n",
    "### Negative:\n",
    "\n",
    "middleton_background=pd.read_table('./Datasets/MiddletonKimBMCBiol2019/bg_pool.no_loc.txt')\n",
    "\n",
    "middleton_background_genes=middleton_background[middleton_background.Stat<-5].Gene.values\n",
    "\n",
    "middleton_background_df=get_wt_utrs(middleton_background_genes) # 340\n",
    "\n",
    "middleton_background_df.to_pickle('./dataframes/design/middleton_background_df.pkl')\n",
    "\n",
    "\n",
    "zappulo_soma=zappulo3[(zappulo3['rzRNA_log2FC_Neurite_Soma']<-2)&(zappulo3.rzRNA_padj_Neurite_Soma<0.01)].gene_name.unique()\n",
    "\n",
    "zappulo_soma_df=get_wt_utrs(zappulo_soma) # 190\n",
    "\n",
    "zappulo_soma_df.to_pickle('./dataframes/design/zappulo_soma_df.pkl')\n",
    "\n",
    "\n",
    "sharedbg=complists(middleton_background_genes,zappulo_soma)\n",
    "\n",
    "# ['Pdp1', 'Rragb', 'St6gal1', 'Kcna6']\n",
    "\n",
    "BurgeCAD=list(zappulo6[zappulo6['Burge_CAD_Neurites.Soma']<-1].index)\n",
    "BurgeN2A=list(zappulo6[zappulo6['Burge_N2A_Neurites.Soma']<-1].index)\n",
    "BurgeCord=list(zappulo6[zappulo6['Burge_Cort_Neurites.Soma']<-1].index)\n",
    "\n",
    "shared_cadn2a=complists(zappulo6.loc[BurgeCAD, 'gene_name'].values, zappulo6.loc[BurgeN2A, 'gene_name'].values)\n",
    "shared_cord=complists(zappulo6.loc[BurgeCord, 'gene_name'].values, shared_cadn2a)\n",
    "\n",
    "# ['Cntn2', 'Disp2', 'Gpr17', 'Magi2', 'Ogt', 'Pgap1']\n",
    "\n",
    "backgroundgenesdf=get_wt_utrs(sharedbg+shared_cord)\n",
    "\n",
    "\n",
    "# selection criteria: UTR > 250 nt, only one isoform, function in the nucleus or cell body\n",
    "\n",
    "backgroundgenes=['Rragb','St6gal1','Gpr17','Ogt','Pgap1']\n",
    "\n",
    "bggenes_df=get_wt_utrs(backgroundgenes)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "overlap backgroundgenes, deDentlist\n",
      "0\n",
      "overlap backgroundgenes, consDentlist\n",
      "0\n",
      "overlap backgroundgenes, isoDentlist\n",
      "0\n",
      "overlap backgroundgenes, zappulo_dend\n",
      "0\n",
      "overlap asoverlap, deDentlist\n",
      "0\n",
      "overlap asoverlap, consDentlist\n",
      "8\n",
      "overlap asoverlap, isoDentlist\n",
      "17\n",
      "overlap asoverlap, zappulo_dend\n",
      "0\n",
      "overlap posgenes, deDentlist\n",
      "4\n",
      "overlap posgenes, consDentlist\n",
      "9\n",
      "overlap posgenes, isoDentlist\n",
      "3\n",
      "overlap posgenes, zappulo_dend\n",
      "0\n",
      "overlap top40, deDentlist\n",
      "12\n",
      "overlap top40, consDentlist\n",
      "37\n",
      "overlap top40, isoDentlist\n",
      "4\n",
      "overlap top40, zappulo_dend\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('overlap backgroundgenes, deDentlist')\n",
    "complists(backgroundgenes, deDentlist)\n",
    "print('overlap backgroundgenes, consDentlist')\n",
    "complists(backgroundgenes, consDentlist)\n",
    "print('overlap backgroundgenes, isoDentlist')\n",
    "complists(backgroundgenes, isoDentlist)\n",
    "print('overlap backgroundgenes, zappulo_dend')\n",
    "complists(backgroundgenes, zappulo_dend)\n",
    "# all 0\n",
    "\n",
    "\n",
    "print('overlap asoverlap, deDentlist')\n",
    "complists(asoverlap, deDentlist)\n",
    "print('overlap asoverlap, consDentlist')\n",
    "complists(asoverlap, consDentlist)\n",
    "print('overlap asoverlap, isoDentlist')\n",
    "complists(asoverlap, isoDentlist) # 8\n",
    "print('overlap asoverlap, zappulo_dend')\n",
    "complists(asoverlap, zappulo_dend)\n",
    "\n",
    "print('overlap posgenes, deDentlist')\n",
    "complists(posgenes, deDentlist)\n",
    "print('overlap posgenes, consDentlist')\n",
    "complists(posgenes, consDentlist)\n",
    "print('overlap posgenes, isoDentlist')\n",
    "complists(posgenes, isoDentlist)\n",
    "print('overlap posgenes, zappulo_dend')\n",
    "complists(posgenes, zappulo_dend)\n",
    "\n",
    "\n",
    "print('overlap top40, deDentlist')\n",
    "complists(top40, deDentlist)\n",
    "print('overlap top40, consDentlist')\n",
    "complists(top40, consDentlist)\n",
    "print('overlap top40, isoDentlist')\n",
    "complists(top40, isoDentlist)\n",
    "print('overlap top40, zappulo_dend')\n",
    "complists(top40, zappulo_dend)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:9: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "  if __name__ == '__main__':\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:29: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:50: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:70: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:100: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n"
     ]
    }
   ],
   "source": [
    "#%% scanning of wild type\n",
    "\n",
    "### posgenes\n",
    "\n",
    "vars_scan50_posgenes=pd.DataFrame()\n",
    "\n",
    "for idx in posgenesdf.index:\n",
    "    seqs=make_scanning(posgenesdf.sequence_minuspa[idx], 50)\n",
    "    genedf = pd.Series(seqs, name='varseq').to_frame()\n",
    "    genedf['gene_name']=posgenesdf.gene_name[idx]\n",
    "    genedf['transcript']=idx\n",
    "    genedf['subset']='wt scanning 50'\n",
    "    genedf['positioninutr']=pd.Series(np.arange(0,50*len(seqs),50))\n",
    "    vars_scan50_posgenes=vars_scan50_posgenes.append(genedf, ignore_index=True)\n",
    "    \n",
    "vars_scan50_posgenes.dropna(inplace=True, subset=['varseq'])\n",
    "vars_scan50_posgenes.drop_duplicates(inplace=True, subset=['varseq'])\n",
    "\n",
    "vars_scan50_posgenes.to_pickle('./dataframes/design/vars_scan50_posgenes.pkl')\n",
    "\n",
    "# 715\n",
    "\n",
    "#### top40\n",
    "\n",
    "vars_scan50_top40=pd.DataFrame()\n",
    "\n",
    "for idx in top40df.index:\n",
    "    seqs=make_scanning(top40df.sequence_minuspa[idx], 50)\n",
    "    genedf = pd.Series(seqs, name='varseq').to_frame()\n",
    "    genedf['gene_name']=top40df.gene_name[idx]\n",
    "    genedf['transcript']=idx\n",
    "    genedf['subset']='wt scanning 50'\n",
    "    genedf['positioninutr']=pd.Series(np.arange(0,50*len(seqs),50))\n",
    "    vars_scan50_top40=vars_scan50_top40.append(genedf, ignore_index=True)\n",
    "    \n",
    "vars_scan50_top40.dropna(inplace=True, subset=['varseq'])\n",
    "vars_scan50_top40.drop_duplicates(inplace=True, subset=['varseq'])\n",
    "\n",
    "vars_scan50_top40.to_pickle('./dataframes/design/vars_scan50_top40.pkl')\n",
    "\n",
    "# 1178\n",
    "\n",
    "\n",
    "#### backgroundgenesdf\n",
    "\n",
    "vars_scan50_backgroundgenes=pd.DataFrame()\n",
    "\n",
    "for idx in backgroundgenesdf.index:\n",
    "    seqs=make_scanning(backgroundgenesdf.sequence_minuspa[idx], 50)\n",
    "    genedf = pd.Series(seqs, name='varseq').to_frame()\n",
    "    genedf['gene_name']=backgroundgenesdf.gene_name[idx]\n",
    "    genedf['transcript']=idx\n",
    "    genedf['subset']='wt scanning 50'\n",
    "    genedf['positioninutr']=pd.Series(np.arange(0,50*len(seqs),50))\n",
    "    vars_scan50_backgroundgenes=vars_scan50_backgroundgenes.append(genedf, ignore_index=True)\n",
    "    \n",
    "vars_scan50_backgroundgenes.dropna(inplace=True, subset=['varseq'])\n",
    "vars_scan50_backgroundgenes.drop_duplicates(inplace=True, subset=['varseq'])\n",
    "\n",
    "vars_scan50_backgroundgenes.to_pickle('./dataframes/design/vars_scan50_backgroundgenes.pkl')\n",
    "\n",
    "# 710\n",
    "\n",
    "#### deDent\n",
    "\n",
    "vars_scan50_deDend=pd.DataFrame()\n",
    "\n",
    "for idx in deDend_df.index:\n",
    "    seqs=make_scanning(deDend_df.sequence_minuspa[idx], 50)\n",
    "    genedf = pd.Series(seqs, name='varseq').to_frame()\n",
    "    genedf['gene_name']=deDend_df.gene_name[idx]\n",
    "    genedf['transcript']=idx\n",
    "    genedf['subset']='wt scanning 50'\n",
    "    genedf['positioninutr']=pd.Series(np.arange(0,50*len(seqs),50))\n",
    "    vars_scan50_deDend=vars_scan50_deDend.append(genedf, ignore_index=True)\n",
    "    \n",
    "vars_scan50_deDend.dropna(inplace=True, subset=['varseq'])\n",
    "vars_scan50_deDend.drop_duplicates(inplace=True, subset=['varseq'])\n",
    "\n",
    "vars_to_drop3=[]\n",
    "for i in vars_scan50_deDend.index:\n",
    "    seq=vars_scan50_deDend.varseq[i]\n",
    "    if ('AATAAA' in seq) | ('ATTAAA' in seq):\n",
    "        vars_to_drop3.append(i)\n",
    "   \n",
    "vars_scan50_deDend.drop(vars_to_drop3, inplace=True)     \n",
    "vars_scan50_deDend.to_pickle('./dataframes/design/vars_scan50_deDend.pkl')\n",
    "\n",
    "\n",
    "# 7948\n",
    "\n",
    "\n",
    "\n",
    "#### zappulo_dend\n",
    "\n",
    "vars_scan50_zappulo_dend=pd.DataFrame()\n",
    "\n",
    "for idx in zappulo_dend_df.index:\n",
    "    seqs=make_scanning(zappulo_dend_df.sequence_minuspa[idx], 50)\n",
    "    genedf = pd.Series(seqs, name='varseq').to_frame()\n",
    "    genedf['gene_name']=zappulo_dend_df.gene_name[idx]\n",
    "    genedf['transcript']=idx\n",
    "    genedf['subset']='wt scanning 50'\n",
    "    genedf['positioninutr']=pd.Series(np.arange(0,50*len(seqs),50))\n",
    "    vars_scan50_zappulo_dend=vars_scan50_zappulo_dend.append(genedf, ignore_index=True)\n",
    "    \n",
    "vars_scan50_zappulo_dend.dropna(inplace=True, subset=['varseq'])\n",
    "vars_scan50_zappulo_dend.drop_duplicates(inplace=True, subset=['varseq'])\n",
    "\n",
    "vars_to_drop4=[]\n",
    "for i in vars_scan50_zappulo_dend.index:\n",
    "    seq=vars_scan50_zappulo_dend.varseq[i]\n",
    "    if ('AATAAA' in seq)|('ATTAAA' in seq):\n",
    "        vars_to_drop4.append(i)\n",
    "   \n",
    "vars_scan50_zappulo_dend.drop(vars_to_drop4, inplace=True)     \n",
    "\n",
    "vars_scan50_zappulo_dend.to_pickle('./dataframes/design/vars_scan50_zappulo_dend.pkl')\n",
    "\n",
    "\n",
    "# 2000\n",
    "\n",
    "\n",
    "#### asoverlap\n",
    "\n",
    "vars_scan50_asoverlap=pd.DataFrame()\n",
    "\n",
    "for idx in asoverlap_df.index:\n",
    "    seqs=make_scanning(asoverlap_df.sequence_minuspa[idx], 50)\n",
    "    genedf = pd.Series(seqs, name='varseq').to_frame()\n",
    "    genedf['gene_name']=asoverlap_df.gene_name[idx]\n",
    "    genedf['transcript']=idx\n",
    "    genedf['subset']='wt scanning 50'\n",
    "    genedf['positioninutr']=pd.Series(np.arange(0,50*len(seqs),50))\n",
    "    vars_scan50_asoverlap=vars_scan50_asoverlap.append(genedf, ignore_index=True)\n",
    "    \n",
    "vars_scan50_asoverlap.dropna(inplace=True, subset=['varseq'])\n",
    "vars_scan50_asoverlap.drop_duplicates(inplace=True, subset=['varseq'])\n",
    "\n",
    "vars_scan50_asoverlap.to_pickle('./dataframes/design/vars_scan50_asoverlap.pkl')\n",
    "\n",
    "# 1203\n",
    "\n",
    "\n",
    "wt_scan = pd.concat([vars_scan50_posgenes, vars_scan50_posgenes[[x in ['Nrgn','Actb','Map2'] for x in vars_scan50_posgenes.gene_name]], \n",
    "                     vars_scan50_top40, vars_scan50_backgroundgenes, \n",
    "                    vars_scan50_deDend, vars_scan50_zappulo_dend, vars_scan50_asoverlap], \n",
    "                    ignore_index=True)\n",
    "\n",
    "wt_scan['changes']='no changes'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "vars_scan50_posgenes: 715\n",
    "vars_scan50_top40: 1178\n",
    "vars_scan50_backgroundgenes: 710\n",
    "vars_scan50_deDend: 7948\n",
    "vars_scan50_zappulo_dend: 2000\n",
    "vars_scan50_asoverlap: 1203\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "4\n",
      "9\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>binding_motif</th>\n",
       "      <th>Pval</th>\n",
       "      <th>Qval</th>\n",
       "      <th>target.hits</th>\n",
       "      <th>bkg.hits</th>\n",
       "      <th>target.seq.hits</th>\n",
       "      <th>bkg.seq.hits</th>\n",
       "      <th>target.seq.count</th>\n",
       "      <th>bkg.seq.count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HNRNPC</td>\n",
       "      <td>AUUUUUA</td>\n",
       "      <td>5.906104e-55</td>\n",
       "      <td>1.771831e-52</td>\n",
       "      <td>5042</td>\n",
       "      <td>2077</td>\n",
       "      <td>533</td>\n",
       "      <td>261</td>\n",
       "      <td>612</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HNRNPCL1</td>\n",
       "      <td>AUUUUUA</td>\n",
       "      <td>5.906104e-55</td>\n",
       "      <td>1.771831e-52</td>\n",
       "      <td>5042</td>\n",
       "      <td>2077</td>\n",
       "      <td>533</td>\n",
       "      <td>261</td>\n",
       "      <td>612</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HuR</td>\n",
       "      <td>UUAAUUU</td>\n",
       "      <td>6.661631e-54</td>\n",
       "      <td>1.985166e-51</td>\n",
       "      <td>4342</td>\n",
       "      <td>1745</td>\n",
       "      <td>523</td>\n",
       "      <td>240</td>\n",
       "      <td>612</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZCRB1</td>\n",
       "      <td>GAUUUAAU</td>\n",
       "      <td>4.601611e-51</td>\n",
       "      <td>1.366678e-48</td>\n",
       "      <td>7198</td>\n",
       "      <td>3182</td>\n",
       "      <td>575</td>\n",
       "      <td>303</td>\n",
       "      <td>612</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>KHDRBS1</td>\n",
       "      <td>UUAAAAA</td>\n",
       "      <td>2.151655e-39</td>\n",
       "      <td>6.347381e-37</td>\n",
       "      <td>4795</td>\n",
       "      <td>2077</td>\n",
       "      <td>532</td>\n",
       "      <td>270</td>\n",
       "      <td>612</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>HNRNPA1L2</td>\n",
       "      <td>UUAGGUA</td>\n",
       "      <td>4.101886e-07</td>\n",
       "      <td>1.633914e-05</td>\n",
       "      <td>1393</td>\n",
       "      <td>660</td>\n",
       "      <td>419</td>\n",
       "      <td>187</td>\n",
       "      <td>612</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>G3BP2</td>\n",
       "      <td>AGGAUAU</td>\n",
       "      <td>2.812262e-06</td>\n",
       "      <td>6.189752e-05</td>\n",
       "      <td>1558</td>\n",
       "      <td>758</td>\n",
       "      <td>452</td>\n",
       "      <td>204</td>\n",
       "      <td>612</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>YBX2</td>\n",
       "      <td>AACAACU</td>\n",
       "      <td>2.920139e-06</td>\n",
       "      <td>6.189752e-05</td>\n",
       "      <td>1702</td>\n",
       "      <td>835</td>\n",
       "      <td>454</td>\n",
       "      <td>226</td>\n",
       "      <td>612</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>SRSF1</td>\n",
       "      <td>AAGGACAU</td>\n",
       "      <td>3.035948e-06</td>\n",
       "      <td>6.189752e-05</td>\n",
       "      <td>5926</td>\n",
       "      <td>3141</td>\n",
       "      <td>583</td>\n",
       "      <td>315</td>\n",
       "      <td>612</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>RBM3</td>\n",
       "      <td>GAUACUA</td>\n",
       "      <td>4.166607e-06</td>\n",
       "      <td>6.189752e-05</td>\n",
       "      <td>1027</td>\n",
       "      <td>481</td>\n",
       "      <td>401</td>\n",
       "      <td>176</td>\n",
       "      <td>612</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein binding_motif          Pval          Qval  target.hits  \\\n",
       "1       HNRNPC       AUUUUUA  5.906104e-55  1.771831e-52         5042   \n",
       "2     HNRNPCL1       AUUUUUA  5.906104e-55  1.771831e-52         5042   \n",
       "3          HuR       UUAAUUU  6.661631e-54  1.985166e-51         4342   \n",
       "4        ZCRB1      GAUUUAAU  4.601611e-51  1.366678e-48         7198   \n",
       "6      KHDRBS1       UUAAAAA  2.151655e-39  6.347381e-37         4795   \n",
       "..         ...           ...           ...           ...          ...   \n",
       "262  HNRNPA1L2       UUAGGUA  4.101886e-07  1.633914e-05         1393   \n",
       "279      G3BP2       AGGAUAU  2.812262e-06  6.189752e-05         1558   \n",
       "281       YBX2       AACAACU  2.920139e-06  6.189752e-05         1702   \n",
       "282      SRSF1      AAGGACAU  3.035948e-06  6.189752e-05         5926   \n",
       "293       RBM3       GAUACUA  4.166607e-06  6.189752e-05         1027   \n",
       "\n",
       "     bkg.hits  target.seq.hits  bkg.seq.hits  target.seq.count  bkg.seq.count  \n",
       "1        2077              533           261               612            377  \n",
       "2        2077              533           261               612            377  \n",
       "3        1745              523           240               612            377  \n",
       "4        3182              575           303               612            377  \n",
       "6        2077              532           270               612            377  \n",
       "..        ...              ...           ...               ...            ...  \n",
       "262       660              419           187               612            377  \n",
       "279       758              452           204               612            377  \n",
       "281       835              454           226               612            377  \n",
       "282      3141              583           315               612            377  \n",
       "293       481              401           176               612            377  \n",
       "\n",
       "[76 rows x 10 columns]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "#%% RBP motifs\n",
    "\n",
    "# from Middleton\n",
    "# SRSF3-binding motif AUCAWCG   Therefore, one hypothesis could be that SRSF3 \n",
    "# plays a role in the early steps of dendritic localization by promoting inclusion \n",
    "# of alternative 3′UTRs (theoretically containing DTEs) and by facilitating nuclear \n",
    "# export.\n",
    "# de novo motifs: UUCGAU CCGCAA GUGGGU\n",
    "\n",
    "dendkmers_middletonpub=['ATCAACG','ATCATCG','TTCGAT','CCGCAA','GTGGGT']\n",
    "\n",
    "# ZBP?\n",
    "\n",
    "### From Taliaferro\n",
    "\n",
    "dendkmers_taliaferropub=['GCTGCT','CTGCTG','GCGCTG','CTGGAC','CCTGCT','TCTGGA','CCCCAA','CTGCCC','ACACTG','TTTTCA','TTTTTT','ATACAG']\n",
    "somakmers_taliaferropub=['TAGGTC','TCTTCT','CTCTTT','TCTCTT','TCTCTC','AGGTAA']\n",
    "\n",
    "\n",
    "\n",
    "### From Ati\n",
    "\n",
    "middleton_dendkmers=pd.read_table('./Datasets/Middleton/deDent_VS_bk_results.txt')\n",
    "middleton_somakmers=pd.read_table('./Datasets/Middleton/bk_VS_deDent_results.txt')\n",
    "\n",
    "dendkmers_middleton = middleton_dendkmers.drop_duplicates('protein',keep='first').drop_duplicates('binding_motif',keep='first').binding_motif.values #28\n",
    "somakmers_middleton = middleton_somakmers.drop_duplicates('protein',keep='first').drop_duplicates('binding_motif',keep='first').binding_motif.values #42\n",
    "\n",
    "zappulo_dendkmers=pd.read_table('./Datasets/Zappulo/neurit_th1_VS_soma_thminus2.txt')\n",
    "zappulo_somakmers=pd.read_table('./Datasets/Zappulo/soma_thminus2_VS_neurit_th1.txt')\n",
    "\n",
    "dendkmers_zappulo = zappulo_dendkmers.drop_duplicates('protein',keep='first').drop_duplicates('binding_motif',keep='first').binding_motif.values #15\n",
    "somakmers_zappulo = zappulo_somakmers.drop_duplicates('protein',keep='first').drop_duplicates('binding_motif',keep='first').binding_motif.values #45\n",
    "\n",
    "\n",
    "complists(dendkmers_middleton, somakmers_middleton) # 0\n",
    "dendkmers_middletonzappulo = complists(dendkmers_middleton, dendkmers_zappulo) # 4\n",
    "somakmers_middletonzappulo = complists(somakmers_middleton, somakmers_zappulo) # 9\n",
    "complists(dendkmers_zappulo, somakmers_zappulo) # 0\n",
    "\n",
    "dendkmers_short = list(set(dendkmers_middletonpub + dendkmers_taliaferropub + [mtf.replace('U','T') for mtf in dendkmers_middletonzappulo]))\n",
    "somakmers_short = list(set(somakmers_taliaferropub + [mtf.replace('U','T') for mtf in somakmers_middletonzappulo]))\n",
    "# 21 and 15 = 36\n",
    "\n",
    "dendkmers_long = list(set([mtf.replace('U','T') for mtf in dendkmers_middleton] + [mtf.replace('U','T') for mtf in dendkmers_zappulo] + dendkmers_middletonpub + dendkmers_middletonpub))\n",
    "somakmers_long = list(set([mtf.replace('U','T') for mtf in somakmers_middleton] + [mtf.replace('U','T') for mtf in somakmers_zappulo] + somakmers_taliaferropub))\n",
    "# 44 and 84 = 128\n",
    "\n",
    "# Get RBPs\n",
    "dendkmers_middletonzappulo_rbps = middleton_dendkmers[[x in dendkmers_middletonzappulo for x in middleton_dendkmers.binding_motif]]\n",
    "'''\n",
    "     protein binding_motif  ...  target.seq.count  bkg.seq.count\n",
    "1      PCBP1      CCCUUCCC  ...               377            612\n",
    "36    SAMD4A       GCUGGCC  ...               377            612\n",
    "42   HNRNPH2       GGGAGGG  ...               377            612\n",
    "43    LIN28A       GGGAGGG  ...               377            612\n",
    "169  BRUNOL6       UGUGGGG  ...               377            612\n",
    "'''\n",
    "middleton_dendkmers[[x in dendkmers_middleton for x in middleton_dendkmers.binding_motif]]\n",
    "'''\n",
    "       protein binding_motif  ...  target.seq.count  bkg.seq.count\n",
    "1        PCBP1      CCCUUCCC  ...               377            612\n",
    "2        PCBP2       CCCUCCC  ...               377            612\n",
    "7       HNRNPK       CCAGCCC  ...               377            612\n",
    "9      HNRNPH2       GGGGGGG  ...               377            612\n",
    "15      PABPC1       AAAAAAA  ...               377            612\n",
    "16      PABPC4       AAAAAAA  ...               377            612\n",
    "17       SART3       AAAAAAA  ...               377            612\n",
    "19       PPRC1       GCGCGCC  ...               377            612\n",
    "20        RBM4       GCGCGCC  ...               377            612\n",
    "21       RBM8A       GCGCGCC  ...               377            612\n",
    "24       PCBP3      CUAUCCCU  ...               377            612\n",
    "28        RBM5       GAGGGAG  ...               377            612\n",
    "29        RBM5       GAGGGAG  ...               377            612\n",
    "30       ESRP2       UGGGGGG  ...               377            612\n",
    "32       SRSF1       GGAGGGC  ...               377            612\n",
    "33       SRSF1       GGAGGGC  ...               377            612\n",
    "34       SRSF1       GGAGGGC  ...               377            612\n",
    "35       SRSF1       GGAGGGC  ...               377            612\n",
    "36      SAMD4A       GCUGGCC  ...               377            612\n",
    "42     HNRNPH2       GGGAGGG  ...               377            612\n",
    "43      LIN28A       GGGAGGG  ...               377            612\n",
    "45      SRSF10       AGAGGGA  ...               377            612\n",
    "55       SRSF2      AGGAGAGG  ...               377            612\n",
    "56      HNRNPL       CCACACA  ...               377            612\n",
    "58      PABPC1       AGAAAAA  ...               377            612\n",
    "59      PABPC4       AGAAAAA  ...               377            612\n",
    "60      PABPC5       AGAAAAA  ...               377            612\n",
    "61       SART3       AGAAAAA  ...               377            612\n",
    "62      SRSF10       AGAAAAA  ...               377            612\n",
    "76     HNRNPH2       GGGAGGC  ...               377            612\n",
    "77       SRSF9       GGGAGGC  ...               377            612\n",
    "84         FUS       CGCGCGC  ...               377            612\n",
    "87     IGF2BP2       AAAAACA  ...               377            612\n",
    "88     IGF2BP3       AAAAACA  ...               377            612\n",
    "89      PABPC3       AAAAACA  ...               377            612\n",
    "90      PABPC4       AAAAACA  ...               377            612\n",
    "96      ZC3H10       CCAGCGC  ...               377            612\n",
    "112       FMR1       GGACGGG  ...               377            612\n",
    "113       FXR2       GGACGGG  ...               377            612\n",
    "114      SRSF1       GGACGGG  ...               377            612\n",
    "136    HNRNPA1       GUAGGGG  ...               377            612\n",
    "137  HNRNPA2B1       GUAGGGG  ...               377            612\n",
    "138       SFPQ       GUGGUGG  ...               377            612\n",
    "145     PABPN1       AAAAGAA  ...               377            612\n",
    "165      CNOT4       GACAGAG  ...               377            612\n",
    "169    BRUNOL6       UGUGGGG  ...               377            612\n",
    "182      RBM38       GUGUGGG  ...               377            612\n",
    "200     HNRPLL       ACACGCA  ...               377            612\n",
    "'''\n",
    "\n",
    "somakmers_middletonzappulo_rbps = middleton_somakmers[[x in somakmers_middletonzappulo for x in middleton_somakmers.binding_motif]]\n",
    "'''\n",
    "      protein binding_motif  ...  target.seq.count  bkg.seq.count\n",
    "1      HNRNPC       AUUUUUA  ...               612            377\n",
    "2    HNRNPCL1       AUUUUUA  ...               612            377\n",
    "3         HuR       UUAAUUU  ...               612            377\n",
    "36       TUT1       AAAUACU  ...               612            377\n",
    "90       SFPQ       GUAAUGU  ...               612            377\n",
    "92        QKI       ACUAAUA  ...               612            377\n",
    "127     RBMS1       AAUAUAC  ...               612            377\n",
    "128     RBMS3       AAUAUAC  ...               612            377\n",
    "134     PCBP3      AUUUUCCU  ...               612            377\n",
    "142    DAZAP1       UAGUUAA  ...               612            377\n",
    "143      MSI1       UAGUUAA  ...               612            377\n",
    "180     RBM28       AAGUAGA  ...               612            377\n",
    "'''\n",
    "middleton_somakmers[[x in somakmers_middleton for x in middleton_somakmers.binding_motif]]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:51: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:58: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:59: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:63: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:64: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:68: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:69: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:73: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:74: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:88: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:95: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:96: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:100: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:101: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:115: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:122: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:123: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:127: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:128: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:156: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:157: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_with_indexer(indexer, value)\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:190: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:207: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:225: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:242: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:261: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:279: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:298: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:332: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:333: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:338: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:339: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:344: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:345: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:350: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:351: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:356: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:357: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:362: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:363: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:396: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:397: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:402: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/martin/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:403: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "###############\n",
    "\n",
    "### Insert in backgrounds\n",
    "\n",
    "forinsertion=posgenesdf.append(bggenes_df)\n",
    "\n",
    "forinsertion.drop_duplicates(inplace=True, subset=['gene_name'])\n",
    "\n",
    "vars_forinsertion=pd.DataFrame()\n",
    "\n",
    "for idx in forinsertion[forinsertion.seqlength<3800].index:\n",
    "    seqs=make_scanning(forinsertion.sequence_minuspa[idx], 150)\n",
    "    genedf = pd.Series(seqs, name='varseq').to_frame()\n",
    "    genedf['gene_name']=forinsertion.gene_name[idx]\n",
    "    genedf['transcript']=idx\n",
    "    genedf['subset']='wt scanning 150'\n",
    "    genedf['changes']='no changes'\n",
    "    genedf['positioninutr']=pd.Series(np.arange(0,150*len(seqs),150))\n",
    "    vars_forinsertion=vars_forinsertion.append(genedf, ignore_index=True)\n",
    "    \n",
    "vars_forinsertion.dropna(inplace=True, subset=['varseq'])\n",
    "vars_forinsertion.drop_duplicates(inplace=True, subset=['varseq'])\n",
    "\n",
    "\n",
    "\n",
    "vars_singlemotif=pd.DataFrame()\n",
    "\n",
    "for i in vars_forinsertion.index:\n",
    "    for mtf in dendkmers_short+somakmers_short:\n",
    "        toadd = vars_forinsertion.loc[i]\n",
    "        sq = toadd['varseq']\n",
    "        toadd['varseq'] = sq[:50] + mtf + sq[50+len(mtf):100] + mtf + sq[100+len(mtf):]\n",
    "        toadd['changes'] = mtf + ' 2 times at position 50 and 100'\n",
    "        vars_singlemotif=vars_singlemotif.append(toadd, ignore_index=True)\n",
    "\n",
    "vars_singlemotif['subset']='motif inserted scanning 150'\n",
    "\n",
    "vars_forinsertion.append(vars_singlemotif, ignore_index=True).to_pickle('./dataframes/design/vars_singlemotif.pkl')           \n",
    "\n",
    "# 6993\n",
    "\n",
    "vars_camk2aforinsertion=vars_forinsertion[vars_forinsertion.gene_name=='Camk2a']\n",
    "# 22\n",
    "vars_camk2aforinsertion['subset']='motifs in Camk2a'\n",
    "\n",
    "\n",
    "vars_combmotif=pd.DataFrame()\n",
    "\n",
    "for i in vars_camk2aforinsertion.index:\n",
    "    toadd = vars_camk2aforinsertion.loc[i]\n",
    "    toadd['changes']='no changes'\n",
    "    vars_combmotif=vars_combmotif.append(toadd, ignore_index=True)\n",
    "\n",
    "    for mtf in dendkmers_short+somakmers_short:\n",
    "        toadd = vars_camk2aforinsertion.loc[i]\n",
    "        sq = toadd['varseq']\n",
    "        newsq = sq[:30] + mtf + sq[30+len(mtf):]\n",
    "        toadd['varseq'] = newsq\n",
    "        toadd['changes'] = mtf + ' 1 time at position 30'\n",
    "        vars_combmotif=vars_combmotif.append(toadd, ignore_index=True)\n",
    "\n",
    "        newsq2 = newsq[:90] + mtf + newsq[90+len(mtf):]\n",
    "        toadd['varseq'] = newsq2\n",
    "        toadd['changes'] = mtf + ' 2 times at position 30,90'\n",
    "        vars_combmotif=vars_combmotif.append(toadd, ignore_index=True)\n",
    "\n",
    "        newsq3 = newsq2[:60] + mtf + newsq2[60+len(mtf):]\n",
    "        toadd['varseq'] = newsq3\n",
    "        toadd['changes'] = mtf + ' 3 times at position 30,60,90'\n",
    "        vars_combmotif=vars_combmotif.append(toadd, ignore_index=True)\n",
    "\n",
    "        newsq4 = newsq3[:120] + mtf + newsq3[120+len(mtf):]\n",
    "        toadd['varseq'] = newsq4\n",
    "        toadd['changes'] = mtf + ' 4 times at position 30,60,90,120'\n",
    "        vars_combmotif=vars_combmotif.append(toadd, ignore_index=True)\n",
    "\n",
    "# 3190\n",
    "vars_combmotif['subset'] = 'motifs inserted, short, Camk2a, up to 4 times'\n",
    "\n",
    "vars_combmotif.to_pickle('./dataframes/design/vars_combmotif.pkl')           \n",
    "\n",
    "###\n",
    "\n",
    "vars_combmotif2=pd.DataFrame()\n",
    "\n",
    "for i in vars_camk2aforinsertion.index:\n",
    "    toadd = vars_camk2aforinsertion.loc[i]\n",
    "    toadd['changes']='no changes'\n",
    "    vars_combmotif2=vars_combmotif2.append(toadd, ignore_index=True)\n",
    "\n",
    "    for mtf in dendkmers_long:\n",
    "        toadd = vars_camk2aforinsertion.loc[i]\n",
    "        sq = toadd['varseq']\n",
    "        newsq = sq[:50] + mtf + sq[50+len(mtf):]\n",
    "        toadd['varseq'] = newsq\n",
    "        toadd['changes'] = mtf + ' at position 50'\n",
    "        vars_combmotif2=vars_combmotif2.append(toadd, ignore_index=True)\n",
    "\n",
    "        newsq2 = newsq[:100] + mtf + newsq[100+len(mtf):]\n",
    "        toadd['varseq'] = newsq2\n",
    "        toadd['changes'] = mtf + ' 2 times at position 50,100'\n",
    "        vars_combmotif2=vars_combmotif2.append(toadd, ignore_index=True)\n",
    "\n",
    "# 1958\n",
    "vars_combmotif2['subset'] = 'motifs inserted, all, Camk2a'\n",
    "\n",
    "vars_combmotif2.to_pickle('./dataframes/design/vars_combmotif2.pkl')           \n",
    "\n",
    "\n",
    "##\n",
    "vars_combmotif2sec=pd.DataFrame()\n",
    "\n",
    "for i in vars_camk2aforinsertion.index:\n",
    "    toadd = vars_camk2aforinsertion.loc[i]\n",
    "    toadd['changes']='no changes'\n",
    "    vars_combmotif2sec=vars_combmotif2sec.append(toadd, ignore_index=True)\n",
    "\n",
    "    for mtf in dendkmers_long:\n",
    "        toadd = vars_camk2aforinsertion.loc[i]\n",
    "        sq = toadd['varseq']\n",
    "        newsq = sq[:50] + mtf + sq[50+len(mtf):50+len(mtf)+3] + str(Seq(sq[38:47]).reverse_complement()) + sq[50+len(mtf) + 3 + 9:]\n",
    "        toadd['varseq'] = newsq\n",
    "        toadd['changes'] = mtf + ' at position 50 in 9 nt hairpin'\n",
    "        vars_combmotif2sec=vars_combmotif2sec.append(toadd, ignore_index=True)\n",
    "\n",
    "        newsq2 = newsq[:100] + mtf + newsq[100+len(mtf):100+len(mtf)+3] + str(Seq(newsq[88:97]).reverse_complement()) + newsq[100+len(mtf) + 3 + 9:]\n",
    "        toadd['varseq'] = newsq2\n",
    "        toadd['changes'] = mtf + ' 2 times at position 50,100 in 9 nt hairpin'\n",
    "        vars_combmotif2sec=vars_combmotif2sec.append(toadd, ignore_index=True)\n",
    "\n",
    "# 1958\n",
    "vars_combmotif2sec['subset'] = 'motifs inserted with hairpin'\n",
    "vars_combmotif2sec.to_pickle('./dataframes/design/vars_combmotif2sec.pkl')           \n",
    "\n",
    "\n",
    "# de novo motifs (Ati)\n",
    "\n",
    "middleton_neurite_denovo=pd.read_excel('./Datasets/middleton_denovo/dedent_denovo.xlsx')\n",
    "middleton_soma_denovo=pd.read_excel('./Datasets/middleton_denovo/bk_denovo.xlsx')\n",
    "\n",
    "zappulo_neurite_denovo=pd.read_excel('./Datasets/zappulo_denovo/neurite_denovo.xlsx')\n",
    "zappulo_soma_denovo=pd.read_excel('./Datasets/zappulo_denovo/soma_denovo.xlsx')\n",
    "\n",
    "denovo_motifs=list(middleton_neurite_denovo.Best_possible_Match.values) + \\\n",
    "    list(middleton_soma_denovo.Best_possible_Match.values) + \\\n",
    "        list(zappulo_neurite_denovo.Best_possible_Match.values) + \\\n",
    "            list(zappulo_soma_denovo.Best_possible_Match.values)\n",
    "\n",
    "\n",
    "vars_denovomotifs=pd.DataFrame()\n",
    "\n",
    "for i in vars_forinsertion[[x in ['Camk2a','Arc','Actb']+backgroundgenes for x in vars_forinsertion.gene_name]].index:\n",
    "    for mtf in denovo_motifs:\n",
    "        toadd = vars_forinsertion.loc[i]\n",
    "        sq = toadd['varseq']\n",
    "        toadd['varseq'] = sq[:30] + mtf + sq[30+len(mtf):]\n",
    "        toadd['changes'] = mtf + ' at position 30'\n",
    "        vars_denovomotifs=vars_denovomotifs.append(toadd, ignore_index=True)\n",
    "\n",
    "# 4554\n",
    "vars_denovomotifs['subset']='de novo motifs'       \n",
    "vars_denovomotifs.to_pickle('./dataframes/design/vars_denovomotifs.pkl')           \n",
    "\n",
    "\n",
    "motifs_comb=pd.concat([vars_forinsertion, vars_singlemotif, vars_combmotif, \n",
    "                       vars_combmotif2, vars_combmotif2sec, vars_denovomotifs], ignore_index=True)\n",
    "\n",
    "# 18653\n",
    "\n",
    "#%%\n",
    "\n",
    "### Mutation of endriched motifs present in variants\n",
    "\n",
    "\n",
    "def scanandreplace(sq, mtf):\n",
    "    sqnew=sq\n",
    "    if sq.find(mtf)>=0:\n",
    "        for startpos in [m.start() for m in re.finditer(mtf, sq)]:\n",
    "            newmtf=''.join([random.choice(['A','C','T','G']) for i in np.arange(len(mtf))])\n",
    "            sqnew=sqnew[:startpos]+newmtf+sqnew[startpos + len(mtf):]\n",
    "    return sqnew\n",
    "\n",
    "vars_motifmut_posgenes = pd.DataFrame()\n",
    "            \n",
    "for i in vars_scan50_posgenes.index:\n",
    "    for j in dendkmers_long:\n",
    "        mutseq=scanandreplace(vars_scan50_posgenes.varseq[i], j)\n",
    "        if mutseq!=vars_scan50_posgenes.varseq[i]:\n",
    "            toadd = vars_scan50_posgenes.loc[i]\n",
    "            toadd.loc['varseq']=mutseq\n",
    "            toadd.loc['changes']=j + ' replaced by random'\n",
    "            \n",
    "            vars_motifmut_posgenes=vars_motifmut_posgenes.append(toadd, ignore_index=True)\n",
    "            \n",
    "# 681\n",
    "            \n",
    "vars_motifmut_posgenes.to_pickle('./dataframes/design/vars_motifmut_posgenes.pkl')           \n",
    "\n",
    "\n",
    "vars_motifmut_top40 = pd.DataFrame()\n",
    "            \n",
    "for i in vars_scan50_top40.index:\n",
    "    for j in dendkmers_long:\n",
    "        mutseq=scanandreplace(vars_scan50_top40.varseq[i], j)\n",
    "        if mutseq!=vars_scan50_top40.varseq[i]:\n",
    "            toadd = vars_scan50_top40.loc[i]\n",
    "            toadd.loc['varseq']=mutseq\n",
    "            toadd.loc['changes']=j + ' replaced by random'\n",
    "            \n",
    "            vars_motifmut_top40=vars_motifmut_top40.append(toadd, ignore_index=True)\n",
    "            \n",
    "# 1010 \n",
    "            \n",
    "vars_motifmut_top40.to_pickle('./dataframes/design/vars_motifmut_top40.pkl')           \n",
    "\n",
    "\n",
    "\n",
    "vars_motifmut_deDend = pd.DataFrame()\n",
    "            \n",
    "for i in vars_scan50_deDend.index:\n",
    "    for j in dendkmers_long:\n",
    "        mutseq=scanandreplace(vars_scan50_deDend.varseq[i], j)\n",
    "        if mutseq!=vars_scan50_deDend.varseq[i]:\n",
    "            toadd = vars_scan50_deDend.loc[i]\n",
    "            toadd.loc['varseq']=mutseq\n",
    "            toadd.loc['changes']=j + ' replaced by random'\n",
    "            \n",
    "            vars_motifmut_deDend=vars_motifmut_deDend.append(toadd, ignore_index=True)\n",
    "            \n",
    "# 6532 \n",
    "            \n",
    "vars_motifmut_deDend.to_pickle('./dataframes/design/vars_motifmut_deDend.pkl')           \n",
    "\n",
    "\n",
    "vars_motifmut_zappulo_dend = pd.DataFrame()\n",
    "            \n",
    "for i in vars_scan50_zappulo_dend.index:\n",
    "    for j in dendkmers_long:\n",
    "        mutseq=scanandreplace(vars_scan50_zappulo_dend.varseq[i], j)\n",
    "        if mutseq!=vars_scan50_zappulo_dend.varseq[i]:\n",
    "            toadd = vars_scan50_zappulo_dend.loc[i]\n",
    "            toadd.loc['varseq']=mutseq\n",
    "            toadd.loc['changes']=j + ' replaced by random'\n",
    "            \n",
    "            vars_motifmut_zappulo_dend=vars_motifmut_zappulo_dend.append(toadd, ignore_index=True)\n",
    "            \n",
    "# 212 \n",
    "            \n",
    "vars_motifmut_zappulo_dend.to_pickle('./dataframes/design/vars_motifmut_zappulo_dend.pkl')           \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "vars_motifmut_asoverlap = pd.DataFrame()\n",
    "            \n",
    "for i in vars_scan50_asoverlap.index:\n",
    "    for j in dendkmers_long:\n",
    "        mutseq=scanandreplace(vars_scan50_asoverlap.varseq[i], j)\n",
    "        if mutseq!=vars_scan50_asoverlap.varseq[i]:\n",
    "            toadd = vars_scan50_asoverlap.loc[i]\n",
    "            toadd.loc['varseq']=mutseq\n",
    "            toadd.loc['changes']=j + ' replaced by random'\n",
    "            \n",
    "            vars_motifmut_asoverlap=vars_motifmut_asoverlap.append(toadd, ignore_index=True)\n",
    "            \n",
    "# 724 \n",
    "            \n",
    "vars_motifmut_asoverlap.to_pickle('./dataframes/design/vars_motifmut_asoverlap.pkl')           \n",
    "\n",
    "\n",
    "\n",
    "vars_motifmut_asoverlapsoma = pd.DataFrame()\n",
    "            \n",
    "for i in vars_scan50_asoverlap.index:\n",
    "    for j in somakmers_long:\n",
    "        mutseq=scanandreplace(vars_scan50_asoverlap.varseq[i], j)\n",
    "        if mutseq!=vars_scan50_asoverlap.varseq[i]:\n",
    "            toadd = vars_scan50_asoverlap.loc[i]\n",
    "            toadd.loc['varseq']=mutseq\n",
    "            toadd.loc['changes']=j + ' replaced by random'\n",
    "            \n",
    "            vars_motifmut_asoverlapsoma=vars_motifmut_asoverlapsoma.append(toadd, ignore_index=True)\n",
    "            \n",
    "# 2745 \n",
    "            \n",
    "vars_motifmut_asoverlapsoma.to_pickle('./dataframes/design/vars_motifmut_asoverlapsoma.pkl')           \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "vars_motifmut_backgroundgenes = pd.DataFrame()\n",
    "            \n",
    "for i in vars_scan50_backgroundgenes.index:\n",
    "    for j in somakmers_long:\n",
    "        mutseq=scanandreplace(vars_scan50_backgroundgenes.varseq[i], j)\n",
    "        if mutseq!=vars_scan50_backgroundgenes.varseq[i]:\n",
    "            toadd = vars_scan50_backgroundgenes.loc[i]\n",
    "            toadd.loc['varseq']=mutseq\n",
    "            toadd.loc['changes']=j + ' replaced by random'\n",
    "            \n",
    "            vars_motifmut_backgroundgenes=vars_motifmut_backgroundgenes.append(toadd, ignore_index=True)\n",
    "            \n",
    "# 1062 \n",
    "            \n",
    "vars_motifmut_backgroundgenes.to_pickle('./dataframes/design/vars_motifmut_backgroundgenes.pkl')           \n",
    "\n",
    "\n",
    "mut_scan = pd.concat([vars_motifmut_posgenes, vars_motifmut_top40, vars_motifmut_backgroundgenes, \n",
    "vars_motifmut_deDend, vars_motifmut_zappulo_dend, vars_motifmut_asoverlap, vars_motifmut_asoverlapsoma], ignore_index=True)\n",
    "\n",
    "mut_scan['subset']='mut scanning 50'\n",
    "\n",
    "# 12967\n",
    "#%% structural motifs\n",
    "\n",
    "# from Middleton: \n",
    "# B1: GAGGCAGGCGGATTTCTGAGTTCGAGGCCAGCCTGGTCTACAGAGTGAGTTCCAGGACAGCCAGGGCTACACAGAGAAACCCTGTCTC\n",
    "# ((((((((....(((((((((((..(((...(((((.((........))..)))))...))).)))))...))))))...))))))))\n",
    "\n",
    "# B2: GCTGGTGAGATGGCTCAGTGGGTAAGAGCACCCGACTGCTCTTCCGAAGGTCAGGAGTTCAAATCCCAGC\n",
    "# (((((.((..((((((....((.(((((((......))))))))).........))).)))..)))))))\n",
    "\n",
    "b1motif='GAGGCAGGCGGATTTCTGAGTTCGAGGCCAGCCTGGTCTACAGAGTGAGTTCCAGGACAGCCAGGGCTACACAGAGAAACCCTGTCTC'\n",
    "b2motif='GCTGGTGAGATGGCTCAGTGGGTAAGAGCACCCGACTGCTCTTCCGAAGGTCAGGAGTTCAAATCCCAGC'\n",
    "zipcode=posgenesdf[posgenesdf.gene_name=='Actb'].sequence.values[0][:54]\n",
    "\n",
    "vars_secmutsbmotifs=pd.DataFrame()\n",
    "\n",
    "for i in vars_forinsertion[[x in ['Camk2a','Arc','Actb']+backgroundgenes for x in vars_forinsertion.gene_name]].index:\n",
    "    toadd = vars_forinsertion.loc[i]\n",
    "    sq = toadd['varseq']\n",
    "    toadd['varseq'] = sq[:30] + b1motif + sq[30+len(b1motif):]\n",
    "    toadd['changes'] = 'B1 motif at position 30'\n",
    "    vars_secmutsbmotifs=vars_secmutsbmotifs.append(toadd, ignore_index=True)\n",
    "    \n",
    "    toadd = vars_forinsertion.loc[i]\n",
    "    sq = toadd['varseq']\n",
    "    toadd['varseq'] = sq[:30] + b2motif + sq[30+len(b2motif):]\n",
    "    toadd['changes'] = 'B2 motif at position 30'\n",
    "    vars_secmutsbmotifs=vars_secmutsbmotifs.append(toadd, ignore_index=True)\n",
    "    \n",
    "    toadd = vars_forinsertion.loc[i]\n",
    "    sq = toadd['varseq']\n",
    "    toadd['varseq'] = zipcode + sq[len(zipcode):]\n",
    "    toadd['changes'] = 'zipcode at position 0'\n",
    "    vars_secmutsbmotifs=vars_secmutsbmotifs.append(toadd, ignore_index=True)\n",
    "    \n",
    "    toadd = vars_forinsertion.loc[i]\n",
    "    sq = toadd['varseq']\n",
    "    toadd['varseq'] = sq[:30] + zipcode + sq[30+len(zipcode):]\n",
    "    toadd['changes'] = 'zipcode at position 30'\n",
    "    vars_secmutsbmotifs=vars_secmutsbmotifs.append(toadd, ignore_index=True)\n",
    "\n",
    "    toadd = vars_forinsertion.loc[i]\n",
    "    sq = toadd['varseq']\n",
    "    toadd['varseq'] = sq[:60] + zipcode + sq[60+len(zipcode):]\n",
    "    toadd['changes'] = 'zipcode at position 60'\n",
    "    vars_secmutsbmotifs=vars_secmutsbmotifs.append(toadd, ignore_index=True)\n",
    "\n",
    "    toadd = vars_forinsertion.loc[i]\n",
    "    sq = toadd['varseq']\n",
    "    toadd['varseq'] = sq[:90] + zipcode + sq[90+len(zipcode):]\n",
    "    toadd['changes'] = 'zipcode at position 90'\n",
    "    vars_secmutsbmotifs=vars_secmutsbmotifs.append(toadd, ignore_index=True)\n",
    "\n",
    "\n",
    "# RNAinverse\n",
    "\n",
    "b1like = ['GGGUUUUAGCCCGAGGUUACUCGUAGCGUGGCUUUUGUAGGAGACGGUAGUGAAAGAAUCGCCCGAGUCGAGAUCUCUCAUAAGAUCC',\n",
    "'CUCUCUCUGAUUACUGGUUAACUAAGUACAUUUCCGACGUCUUGAUCCGGACGGAGCCCUAUAGGUUAAUUACCGGUCCAAGAGGGAG',\t\t\n",
    "'UAUAAGCGCUUCCAUCCGCCGACGCGAAUUCUAUAACGGCCGUCGACCCACUUGUGCUUUUCAGUCGGACGUGGGUGCCACGUUUGUA',\t\n",
    "'UACGGGGCAAUGGCUUGUUCCACCAUUAGGAGGGGUCUGAGCAAAUACAAAAUCCCGGAUAGAGUGGAUAGGCAGGCGUGGUUUCGUA',\t\t\n",
    "'GACAUAAAGAGUCAGUUAGUGGUCAGCACGAUUCUGCGCUUUAAUAUGCCACAGAAGGAUGCCACUACAUCUAGUUGUAAUUUGUGUC',\t\t\n",
    "'GUGACGACGCCUUUUUGGCCGGCCUGUCUGGAUCUAGGCAAUCGCAGGCAAUGGAUAGUGGCUGUUGGACAUCGGAAUCAGUCGUUGC',\t\t\n",
    "'GAAUCACUUAUUCAUAAGUGUUGGCUGCGAACCUUUGUUGGAGGUUGAAUGGAAGGGAAGCGAUAACGAGUUUUAUGCCUAGUGGUUU',\t\t\n",
    "'GGGGGAAGGAGGUGGCGUUGUGGCUUACCACUAAACCCUUGAUGUCUAGCCGUUUGAAAGUAUCCGCAAUCACGCUAGAGCUUCUCUC',\t\t\n",
    "'AGCGAGUUAACGAUCCUAAUUCAGCGGUUUAUUCGUCGGAUCUGAGGCCAAACGGAAGUAUUAUGAGUUCCUAGGGUGACAACUUGCU',\t\t\n",
    "'GUAGGGACCGGGCUUAUGCUCGUACAACAGACCCUUAAGAUUGCACGCUGGAAGGGGACGUUCGCGAGACACGUAAGGGAGUCCCUAC']\t\t\n",
    "\n",
    "b2like = ['CCGGCUGCAUUUAUGACCAUGGAUCCGCGUGCGGGGACGCGGACCUCGAAAUUUUCACUAACGGUGUCGG',\t\t\n",
    "'CCGCAGCGAAUCGCAGUAUAAGCAGCGAUAACUGAAUGUCGCUCUGAGGUCCGUCUGAUGAGGUGUGUGG',\t\t\n",
    "'GGAUAACUCACGCCUAAAAUAGUGGAAGGUUGCGUUGCCUUCCCUCAAUCGAUGUGGUGCGCCAGUGUCC',\t\t\n",
    "'CUGUAAGUAUGCGUCCUAAACCAUUAGUUAGCAUUAUGACUGGGGCACACGGGUGGACCGCCGACUACGG',\t\t\n",
    "'GCUAUAGCACAUCAUCAAGAGAUGUAUUGCGACUUAGUAGUAUUCGGGUAAAGAGAUAGAUAGGCAUAGU',\t\t\n",
    "'GAGAUUAGACCACGCUCUCUUCAUAAGCGAUCAACUUUGCUUGGAUCUUUACACGGCAGUGCGCUGUCUC',\t\t\n",
    "'AGUUGGCUGUGCCAAUUCUUGCAGCCACUAACGACGUGGUGGCGCUUUUCACCUGUUCGGCUAAGCAGCU',\t\t\n",
    "'GUUUAACUAUCACAUCAUUUCUCGAGAUGGAGGUGCCCAUUUCAGUGUCUGUUCGAUCGUGUGAGUGAGC',\t\t\n",
    "'GGACGUAGGAAUUCCAAAAAGUAGAUUCCACUCGAUUGGAGUCACGUAUAAGUAUGGUAGUGGCUCGUCU',\t\t\n",
    "'GACGACACGGGCAUUGAAUAGUAUGUUUUGCUAAUAUAAGACGACAAGGAGGAGCGAUUGCGAGUUUGUC']\n",
    "\n",
    "\n",
    "for i in vars_forinsertion[[x in ['Camk2a','Arc','Actb']+backgroundgenes for x in vars_forinsertion.gene_name]].index:\n",
    "    for mtf in b1like:\n",
    "        toadd = vars_forinsertion.loc[i]\n",
    "        sq = toadd['varseq']\n",
    "        toadd['varseq'] = sq[:30] + mtf.replace('U','T') + sq[30+len(mtf):]\n",
    "        toadd['changes'] = 'B1 like motif at position 30'\n",
    "        vars_secmutsbmotifs=vars_secmutsbmotifs.append(toadd, ignore_index=True)\n",
    "    for mtf in b2like:\n",
    "        toadd = vars_forinsertion.loc[i]\n",
    "        sq = toadd['varseq']\n",
    "        toadd['varseq'] = sq[:30] + mtf.replace('U','T') + sq[30+len(mtf):]\n",
    "        toadd['changes'] = 'B2 like motif at position 30'\n",
    "        vars_secmutsbmotifs=vars_secmutsbmotifs.append(toadd, ignore_index=True)\n",
    "\n",
    "# 2139\n",
    "vars_secmutsbmotifs['subset']='sec structure and zipcode'        \n",
    "vars_secmutsbmotifs.to_pickle('./dataframes/design/vars_secmutsbmotifs.pkl')           \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#%% assemble library\n",
    "\n",
    "library_varseq = pd.concat([wt_scan, mut_scan, motifs_comb, vars_secmutsbmotifs], ignore_index=True)\n",
    "\n",
    "library_varseq.to_pickle('./dataframes/design/library_varseq.pkl')           \n",
    "\n",
    "from Bio import Restriction\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "vars_to_drop=[]\n",
    "for i in library_varseq.index:\n",
    "    seq=library_varseq.varseq[i]\n",
    "    if (len(Restriction.SbfI.search(Seq(seq)))>0) | (len(Restriction.AscI.search(Seq(seq)))>0):\n",
    "        vars_to_drop.append(i)\n",
    "\n",
    "library_varseq.drop(vars_to_drop, inplace=True)\n",
    "\n",
    "\n",
    "#%%\n",
    "\n",
    "bcspl=pd.read_csv('./Datasets/40000barcodes.csv', header=None)\n",
    "bcspl.columns=['idx','barcode']\n",
    "\n",
    "bcs=list(bcspl.barcode.values)\n",
    "bcs2=list(bcspl.barcode.values)\n",
    "\n",
    "\n",
    "primers=['GCCCCACGGAGGTGCCAC','CCTCCTCACGGCGACGCG','CTCCCGGGCATGCGAATT','TCAACCAGTCGCGGTCCA',\n",
    "         'TGCGAGTTAGGGGACGGT','ACGGACGCGGGTATAGCA','CGAAATGGGCCGCATTGC','CACTGCGGCTGATGACGA',\n",
    "            'GACAGATGCGCCGTGGAT','AGCCACCCGATCCAATGC','ATGGGGTTCGGTATGCGC','AAGGCTCCCCGAGACGAT']\n",
    "\n",
    "\n",
    "'''\n",
    "\n",
    "wt scanning 50 CGAAATGGGCCGCATTGC CACTGCGGCTGATGACGA\n",
    "[the rest] GACAGATGCGCCGTGGAT AGCCACCCGATCCAATGC\n",
    "\n",
    "three TGCGAGTTAGGGGACGGT ACGGACGCGGGTATAGCA\n",
    "\n",
    "'''\n",
    "def addbc(idx):\n",
    "    if library_varseq.subset[idx]=='wt scanning 50':\n",
    "        while True:\n",
    "            bc=random.choice(bcs)\n",
    "            vsnew=primers[6] + bc + library_varseq.varseq[idx] + primers[7]\n",
    "            if (Restriction.SbfI.search(Seq(vsnew))==[])&(Restriction.AscI.search(Seq(vsnew))==[]):\n",
    "                bcs.remove(bc)\n",
    "                break\n",
    "    else:\n",
    "        while True:\n",
    "            bc=random.choice(bcs2)\n",
    "            vsnew=primers[8] + bc + library_varseq.varseq[idx] + primers[9]\n",
    "            if (Restriction.SbfI.search(Seq(vsnew))==[])&(Restriction.AscI.search(Seq(vsnew))==[]):\n",
    "                bcs2.remove(bc)\n",
    "                break\n",
    "    return vsnew\n",
    "\n",
    "\n",
    "library_varseq['varseq_final']=library_varseq.index.map(lambda x: addbc(x))\n",
    "\n",
    "\n",
    "\n",
    "vars_to_drop=[]\n",
    "for i in library_varseq.index:\n",
    "    seq=library_varseq.varseq_final[i]\n",
    "    if (len(Restriction.SbfI.search(Seq(seq)))>0) | (len(Restriction.AscI.search(Seq(seq)))>0):\n",
    "        vars_to_drop.append(i)\n",
    "\n",
    "\n",
    "lengths=library_varseq.varseq_final.apply(lambda x: len(x))   \n",
    "\n",
    "\n",
    "library_varseq['barcode']=library_varseq.varseq_final.apply(lambda x: x[16:30])\n",
    "library_varseq['primer1']=library_varseq.varseq_final.apply(lambda x: x[:18])\n",
    "library_varseq['primer2']=library_varseq.varseq_final.apply(lambda x: x[-18:])\n",
    "\n",
    "library_varseq.to_pickle('./dataframes/design/library_varseq_FINAL.pkl')           \n",
    "\n",
    "library_varseq.to_csv('./dataframes/design/library_varseq_FINAL.csv')           \n",
    "\n",
    "library_varseq.varseq_final.to_csv('./dataframes/design/RNAlocneurons_library.csv', header=None)           \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    }
   ],
   "source": [
    "# make sure that there are no other characters than actg\n",
    "\n",
    "noactg=[]\n",
    "count=0\n",
    "for i in library_varseq.varseq_final.values:\n",
    "    for j in i:\n",
    "        if (j!='A')&(j!='T')&(j!='C')&(j!='G'):\n",
    "            noactg.append(i)\n",
    "    count+=1\n",
    "    \n",
    "print(noactg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
